### DISCUSSION TITLE
how to use ref.listener's previous and next value?

### DISCUSSION THREAD

**Discussion by Shvet** [rrousselGit/riverpod#969] [Q&A] [answered]
I was user `flutter_riverpod's` 0.18.0 version in which `ref.listener` was only providing current state of `StateNotifier` after migrating it to 1.0.0. it now provides previous and next state which means if i am listening to statenotifier right now it will give me 2 values which neither confirms my current state of statenotifier. For example i am trying to get data from server like below:

### StateNotifier code:

```
final mainCategoryViewModel = StateNotifierProvider.autoDispose<ParentCategoryNotifier, ParentCategoryState>((ref) {
  return ParentCategoryNotifier(ref.read);
});

class ParentCategoryNotifier extends StateNotifier<ParentCategoryState> {
  final Reader _reader;
  late final ProductsRepositoryImpl _impl = _reader(productRepositoryImplProvider);

  ParentCategoryNotifier(this._reader) : super(ParentCategoryInit()) {
    getAllProductCategory(0);
  }

  Future<void> getAllProductCategory(int parent) async {
    state = ParentCategoryLoading();
    _impl.getAllProductCategory(parent).then((value) => {
          if (value.isSuccess)
            {
              value.ifSuccess((data) => state = ParentCategoryLoaded(data)),
            }
          else if (value.isFailure)
            {
              value.ifFailure(
                (e) => state = ParentCategoryError(
                  UserSignInFailed.fromJson(e.response as Map<String, dynamic>),
                ),
              )
            }
        });
  }
}

```

In above code i am trying to put state for loading, error and success which will be used to show loading dialog for loading state, snackbar for error state and list of data for success state in previous version. Now it doesn't work as i will get `ParentCategoryLoading` in listener's previous and `ParentCategoryLoaded` for next state.

### Listener code:

```
ref.listen<ParentCategoryState>(mainCategoryViewModel, (previous, next) {
      log("$next && $previous");
      if (previous is ParentCategoryLoading) {
        showWaitingDialog(context);
      } else if (next is ParentCategoryLoaded) {
        if (Navigator.canPop(alertBuildContext)) {
          Navigator.pop(alertBuildContext);
        }
        setState(() {
          mainCategory.addAll(next.productCategory);
        });
      } else if (next is ParentCategoryError) {
        if (Navigator.canPop(alertBuildContext)) {
          Navigator.pop(alertBuildContext);
        }
        ScaffoldMessenger.maybeOf(context)!.showSnackBar(SnackBar(
          content: Text(
            "${next.failed.message}",
            style: Theme.of(context).textTheme.bodyText2!.copyWith(
                  color: Colors.white,
                  fontWeight: FontWeight.w400,
                  fontStyle: FontStyle.normal,
                  fontSize: 15.0,
                  letterSpacing: 0.15,
                  height: 1.2,
                ),
            softWrap: true,
          ),
          behavior: SnackBarBehavior.fixed,
          duration: const Duration(minutes: 1),
          elevation: 4.0,
          action: SnackBarAction(
              label: "Ok",
              onPressed: () {
                ScaffoldMessenger.maybeOf(context)!.hideCurrentSnackBar();
              }),
        ));
      }
    });
```

Here what will happen is i will not get data in my listview as well as my dialogbox will keep showing. if i change above code to only listen for next state, i wont be able to show dialogbox as in next state it will never get `ParentCategoryLoading` state.

So i would like to know if there is any example to use in my case for solve this problem?

**Comment by rrousselGit** [comment, marked as answer]
You do want to listen only to "next"
Your issue is, by the time you call `ref.listen`, your notifier is already in loading state, so you won't receive the loading event.

You can solve that by using a ConsumerStatefulWidget + initState by doing something like:

```dart
initState() {
  super.initState();
  if (ref.read(mainCategoryViewModel) is ParentCategoryLoading) {
    Future(() => showWaitingDialog(context));
  }
}
```

**Reply by Shvet** [reply]
Thank it worked.

Actually i was already using `ref.read<ParentCategoryState>(mainCategoryViewModel)` in initState method but didnt know that i can compare there directly.
